Region list page map tab

hudson-newey opened the pull request:
# Project & Region list page map tab

## Changes

- Adds "tiles" and "map" tab to the project & region list pages
- Adds `filter` property to the `baw-site-map` component
- Refactors project & region list page logic into common template
- Adds clustering to Google maps
- Add `isFetching` state to site map component so that the loading spinner can show during data fetching
- Add `loading` class to `baw-site-map` component host that can be used to conditionally apply styles depending on if the site map is loaded or not
	- I use this on the project/region list pages to add a shadow once the site map has loaded. I don't add the shadow while the map is loading because it's too notable
- No longer fetch sites in a site-map component. I did this for two reasons
	- Some website-level site filter requests take a long time, and I don't want to hang SSR for just these requests
	- Site maps are heavily reliant on user authentication state, and return an empty array a lot of the time (because the unauthenticated SSR request wouldn't have access to any of the sites that would be shown). Meaning the map would flash a "No sites" message.
	- I recognize that we might be able to enable this once SSR does authenticated requests in an associated TODO comment
- site-map.component will now enter a `isFetching` state after 400ms of filter conditions changing. This means that we don't flash a loading spinner for fast requests. This makes map updates feel more instant & produces less jank
- Reduces `defaultDebounceTime` to 200ms (from `500ms`). Once we merge this PR, let me know if this results in a lot of server load, and I can increase/revert this change.
- Adds 5px to the annotation import preview table's "Start Time" and "End Time" columns. (I did this because when I imported Phil's annotations, I saw that long start/end times overflowed the allocated cells
- Adds `groupBy` attribute to `site-map.component`

### Development concerns

- Adds `readonly` annotation to signals that were added before I knew about this [recommendation](https://angular.dev/style-guide#use-readonly-for-properties-that-shouldnt-change)
- Fixes `templateUrl` and `styleUrl`s to conform to the Angular style guide [recommendation](https://v19.angular.dev/style-guide#style-05-04)
- Migrate from deprecated `MapMarker` to `AdvancedMapMarker` ([more info](https://mapsplatform.google.com/resources/blog/announcing-highly-customizable-higher-performance-markers/))

## Issues

Fixes: #2436

## Visual Changes

<img width="1337" height="964" alt="image" src="https://github.com/user-attachments/assets/38c2b463-5b64-4702-bf7a-9b165a437b79" />

_Region "tiles" tab (showing A2O)_

Note: Images arn't loading because of CORP. Images do correctly load when on the correct domain

---

<img width="1357" height="929" alt="image" src="https://github.com/user-attachments/assets/af69ecb9-666c-48b9-82e9-dbe5b74870bb" />

_Region "map" tab (showing A2O)_

---

<img width="1378" height="969" alt="image" src="https://github.com/user-attachments/assets/8f4efc8f-2a77-4ccd-96a5-bb7ecf631f8f" />

_Project list page with `homePageAndProjectListMaps: false`_

This is how Ecosounds will look until we add the site clustering endpoint

---

<img width="1074" height="810" alt="image" src="https://github.com/user-attachments/assets/5080e270-c185-42de-a59f-bfc8f009e6ee" />

_New marker colors depending on what region the marker belongs to_

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Region list page map tab

hudson-newey committed:
Marker clustering + text searching

hudson-newey committed:
Support filtering by region name

hudson-newey committed:
Transition to "load" state if map takes > 300ms

For fast loading maps, we don't enter a loading state because it'd flash
white, giving a jarring experience.

For slow requests, we still want to provide feedback to the user.

Note that we never used to have a "data fetching" state. This is a new
feature that I added because waiting 2 seconds for the a2o map to load
after typing in a site name looked really slow.
Note that if you type in a site name that narrows to a small subset of
sites, we never transition into this fetching state.

hudson-newey committed:
Refactor out specific region components

hudson-newey committed:
Shared ModelListComponent

hudson-newey committed:
project list + turn ModelList into component

hudson-newey commented on src/app/components/shared/map/map.component.scss:
```diff
@@ -4,7 +4,7 @@
   align-items: center;
 
-  border: 1px solid black;
+  box-shadow: inset 0 0 0.2rem rgba(0, 0, 0, 0.2);
```
I did this so that rounding could be applied to the map component.

This obviously isn't the correct solution. I'll find a correct solution before merging this PR.

hudson-newey commented on src/app/services/maps/maps.service.ts:
```diff
@@ -118,6 +126,11 @@ export class MapsService {
 
+  private googleMapClusterUrl(): string {
+    // https://github.com/angular/components/blob/974d42f04/src/google-maps/map-marker-clusterer/README.md?plain=1#L5
+    return "https://unpkg.com/@googlemaps/markerclusterer/dist/index.min.js";
```
Before merging, I'll add this as a dependency & static asset that will be served through our Angular server.

This would have two benefits.

1. The version would be tracked in our package.json
2. Hopefully we can re-use the same `Keep-Alive` connection so that we don't have to do another TLS + TCP handshake

hudson-newey committed:
Self-review

hudson-newey committed:
use load state before Google maps is initalized

hudson-newey commented on src/app/services/maps/maps.service.ts:
```diff
@@ -11,7 +11,7 @@ export enum GoogleMapsState {
 }
 
-export type MapMarkerOptions = Partial<google.maps.marker.AdvancedMarkerElement>;
+export type MapMarkerOptions = google.maps.marker.AdvancedMarkerElement;
```
The Partial model was incorrect.

I now correctly initialize `MapMarkerOptions` in the `Site` model using the `AdvancedMarkerElement` constructor

hudson-newey committed:
Copilot feedback

hudson-newey committed:
Serve @googlemaps/markerclusterer with angular SSR

By serving this dependency through the Angular server, we don't have to
establish another connection, and can hopefully serve this dependency
through the existing keep-alive connection.

[skip ci]

atruskie commented on src/app/components/projects/components/site-map/site-map.component.ts:
```diff
@@ -72,19 +87,35 @@ export class SiteMapComponent extends withUnsubscribe() implements OnChanges {
+    // For fast requests, we don't want to flash a loading state.
+    // However, if fetching takes a long time, we do want to show a loading
+    // state to inform the user that something is happening.
+    const timerSub = timer(400)
```
name the magic number. Maybe also put it next to our default debounce const?

hudson-newey replied:
I like the idea

atruskie commented on src/app/components/projects/components/site-map/site-map.component.ts:
```diff
@@ -28,23 +30,29 @@ import { first, takeUntil } from "rxjs/operators";
   public readonly regions = input<IdOr<Region>[]>();
   public readonly sites = input<IdOr<Site>[]>();
   public readonly selected = input<List<IdOr<Site>>>();
+  public readonly filters = input<InnerFilter<Site>>();
```
I'm not overly keen on passing filters through - see our previous discussion on just providing an observable

atruskie replied:
also how do we handle filtering by regoin name?

on the region list page the filter would be `name eq "abc"`, but to pass it into the map, it would need to be be `regions.name eq "abc"`. I'll probably work out where this translation happens elsewhere....

atruskie replied:
Maybe this is just a lack of documentation for the input? I'm making assumptions about it's behaviour...

hudson-newey replied:
> also how do we handle filtering by regoin name?

You're correct, you just pass a `{ "regions.name": { "contains": "abc" } }` object into this input

hudson-newey replied:
> see our previous discussion on just providing an observable

I'm quite opposed to this, why would we want an observable here?

We don't need the functionality of observable's (history/queue of items), and the pagination template doesn't omit anything like an `Observable<Filter>`

atruskie replied:
I think the best idea I had for rearchitecting was in https://github.com/QutEcoacoustics/workbench-client/pull/2443#discussion_r2396408344 and thus my actual suggestion is to remove the filter completely

atruskie commented on src/app/components/shared/model-list/model-list.component.ts:
```diff
@@ -0,0 +1,117 @@
+      // I need to type cast the filter here because our filter types don't
+      // currently reflect that we can filter by associated models.
+      this.mapFilter.set({
+        [`${this.modelKey()}.name`]: { contains: textFilter },
```
this is a bit icky - it will quickly become unmaintainable if we add any more filter criteria.

I wonder if this is better setup as a two-stage filter. Filter for model with search criteria, and then use found ids to return sites (or just feed model ids into baw-site-map directly).

It will also allow us reuse a model filter request (e.g. filter regions) directly in the map without rerunning a request (e.g. filter sites where regions.name = region name) as soon as we turn on the map

hudson-newey replied:
I like the idea of re-using region ids in the site map component, but I think we can only do this when there is a single page of results.

The "tile" view only returns 25 items at a time, meaning that if we pass the region ids returned by the pagination filter request, we'll only show the sites from the top 25 regions.

A possible fix would be to query all projects/regions on the list page. Potentially by still doing paging requests so that we can render the tile view after the first response. But this seems like a lot of overhead / really inefficient to (essentially) keep the entire projects/regions table in the clients memory.

hudson-newey replied:
Discussed offline.

The main problem with this solution is that we'll have to maintain two filters, one for the project/region list page, and one for the site filters.

We came up with a middle ground solution where we add a `associationModelFilter` helper that can merge an association filter (project/region filter) into an existing base filter (sites).

That way we can re-use the same project/region filters and just call this helper to create a site filter.

atruskie commented on src/app/helpers/app-initializer/app-initializer.ts:
```diff
@@ -49,6 +49,7 @@ export interface Settings {
     harvestFilenameGuide: string;
   };
   hideProjects: boolean;
+  websiteMaps: boolean;
```
nit naming: suggest `globalMaps`

hudson-newey replied:
Nit: "global" sounds geographical

if we change this name to `globalMaps`, it might be confusing because you can still have global geographical map with this environment setting turned off.

By a similar thought train, I also originally rejected `siteMaps` because the term "site" was overloaded with "region/site/point" terminology (which is why I originally landed on "websiteMaps").

I'm happy to keep `globalMaps`, just want to make a note that it might be confusing

atruskie replied:
Fair. Happy with a different name. Even long or obtuse - hopefully this is just a temporary config.

`disableHomePageAndProjectListMaps`

hudson-newey replied:
`disableHomePageAndProjectListMaps` This works for me

I'll probably remove the negation and make it something like `homePageAndProjectListMaps`

atruskie commented on src/app/services/config/configMock.service.ts:
```diff
@@ -60,6 +60,7 @@ export const testApiConfig = new Configuration({
       harvestFilenameGuide: "http://broken_link",
     },
     hideProjects: false,
+    websiteMaps: true,
```
if name change, then here too

atruskie commented on src/app/app.helper.ts:
```diff
@@ -3,8 +3,12 @@ import { inject, InjectionToken, PLATFORM_ID } from "@angular/core";
+ *
+ * I chose a 200ms debounce delay because it means that worst case scenario,
+ * we are sending 5 requests per second to the server, which _should_ not
+ * overload the server.
```
worth mentioning this was originally chosen to reduce jank for a user based on typing speed so it's a balance to tune

atruskie reviewed (changes requested):
Great work!

The clustering is going to be necessary, but I'm not fond of the default styles. That dark blue colour has low contrast against the ocean, and it's a real jarring shift up to the bright red colour for larger clusters.


I'm kinda realising that always showing sites in a map for regions/projgects pages is going to be confusing. We didn't get round to coloring markers by group (region/project), did we?

hudson-newey committed:
Review feedback

hudson-newey committed:
Untested associationModelFilter helper

[skip ci]

hudson-newey committed:
Add associationModelFilter tests

[skip ci]

hudson-newey committed:
Revert AdvancedMapMarker

It broke on hover tooltips, and it's not strictly needed for this PR

[skip ci]

hudson-newey committed:
Region + project list tests

hudson-newey committed:
Basic marker grouping

hudson-newey committed:
Use markers for grouping

[skip ci]

hudson-newey committed:
Adjust tests

hudson-newey committed:
Fix hover tooltips on AdvancedMapMarkers

[skip ci]

hudson-newey committed:
Fix failing home tests

[skip ci]

hudson-newey committed:
More advanced MockAdvancedMarkerElement

hudson-newey committed:
Passing test suite

hudson-newey committed:
Merge branch 'master' into list-maps

hudson-newey committed:
model-list.component tests

hudson-newey committed:
Basic group assignment tests

hudson-newey committed:
group color tests

This is the last of the tests. I think I am finished with this PR.
I'll do a self-review and hopefully get to merging this thing.

hudson-newey committed:
Fix indefinite fetching state for short circut

We used to only leave the "fetching" state once the API call completed.
However, because we can sometimes short-circut without an API call, this
caused some issues.

hudson-newey committed:
Revert marker cache

I thought that I could cache marker content between groups, so that each
marker grouping could share an element reference for content (so that we
wouldn't re-create heaps of duplicate pin elements).

However, this didn't seem to work, and only one pin in the group would
have the correct styling.

I still like the idea, so I added a TODO.

However, in the spirit of merging this PR, I'm happy to merge without
pin content caching, because it wasn't really slow to begin with.

hudson-newey committed:
Explicitly hide query qsp in empty pagination filter

hudson-newey committed:
Update pagination tests

hudson-newey committed:
Improved location input

hudson-newey committed:
working location-input.component tests

hudson-newey committed:
Adapt site-map tests

hudson-newey committed:
Simpler location input tests

hudson-newey committed:
Adjust site.component tests

We now wait for the google namespace to be initalized before creating
markers, meaning that the async queue mocks were broken during tests.

hudson-newey committed:
More mock async queue corrections
